🚀 1단계 - 엔티티 매핑

daehyeon0 opened the pull request:
안녕하세요. 리뷰어님.
JPA 첫번째 미션 리뷰 요청드립니다. :)

daehyeon0 committed:
docs(README): 요구사항 추가

daehyeon0 committed:
chore: sql log 볼 수 있도록 설정 변경

daehyeon0 committed:
build: Application 실행 가능한 수준으로 수정

- Answer,DeleteHistory, Question, User Entity 생성

daehyeon0 committed:
feat: User Entity 매핑

daehyeon0 committed:
feat: Question Entity 매핑

daehyeon0 committed:
feat: DeleteHistory Entity 매핑

daehyeon0 committed:
fix: DeleteHistory의 contentType에 Enumberated(EnumType.STRING) 적용

daehyeon0 committed:
feat: User Entity의 userId를 unique로 설정

daehyeon0 committed:
feat: Answer Entity 매핑

daehyeon0 committed:
test: UserTest의 findByUserId -> findById로 변경

daehyeon0 committed:
test: QuestionTest 수정

- save 수정
- findById 추가

daehyeon0 committed:
test: DeleteHistoryTest

- 테스트 간소화

daehyeon0 committed:
test(QuestionTest, UserTest): static data의 id가 아닌 실제 저장된 데이터의 ID를 이용하여 findById

daehyeon0 committed:
test: 잘못된 테스트 수정

- QuestionTest로 쓰일 테스트가 DeleteHistoryTest에 있던 것 수정

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,128 @@
+	private Long id;
+	@Lob
+	private String contents;
+	@Column(nullable = false)
+	private LocalDateTime createdAt;
```
createdAt와 updatedAt는 다른 entity에도 공통으로 사용되는 것 같아요!
`@MappedSuperclass`를 이용해서 중복을 제거해보면 어떨까요?

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,128 @@
 
-import java.util.Objects;
-
+/**
+ * create table answer
+ * (
+ *     id          bigint generated by default as identity,
+ *     contents    clob,
+ *     created_at  timestamp not null,
+ *     deleted     boolean   not null,
+ *     question_id bigint,
+ *     updated_at  timestamp,
+ *     writer_id   bigint,
+ *     primary key (id)
+ * )
+ */
```
주석으로 테이블 스팩을 적어주셔서 저도 리뷰하기 편했습니다 🙇‍♂️ 감사드립니다.

다만 아무래도 이런 주석은 실제 운영 단계에서는 배포되면 안될 것입니다. 다음 단계에서는 삭제하여주시길 부탁드릴께요! 🙏

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,128 @@
+		this.updatedAt = createdAt;
+	}
+
+	protected Answer() {
+	}
```
접근 제한자도 잘 설정해주셨습니다 👍

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,93 +1,128 @@
+
+	public Long getId() {
+		return id;
+	}
+
+	public void setId(Long id) {
+		this.id = id;
+	}
+
+	public Long getWriterId() {
+		return writerId;
+	}
+
+	public void setWriterId(Long writerId) {
+		this.writerId = writerId;
+	}
+
+	public Long getQuestionId() {
+		return questionId;
+	}
+
+	public void setQuestionId(Long questionId) {
+		this.questionId = questionId;
+	}
+
+	public String getContents() {
+		return contents;
+	}
+
+	public void setContents(String contents) {
+		this.contents = contents;
+	}
+
+	public boolean isDeleted() {
+		return deleted;
+	}
+
+	public void setDeleted(boolean deleted) {
+		this.deleted = deleted;
+	}
```
기본으로 생성되어 있는 getter와 setter는 우선 삭제를 권장드리며,
반드시 필요한 경우에만 도입을 검토하는 편을 추천드립니다.

na-mul commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,32 @@
+        Question expected = questionRepository.save(Q2);
+        Question actual = questionRepository.findById(expected.getId()).get();
+
+        assertThat(expected).isEqualTo(actual);
+        assertThat(expected).isSameAs(actual);
```
`isSameAs`는 `isEqualTo` 보다 더욱 엄밀한 의미의 비교입니다. 
equals 메소드를 재정의한다 하더라도, 객체의 주소가 다르다면 `isSameAs`는 false가 될 것입니다. 
현재의 테스트에서 사용하기에는 너무 엄격해보이니 `isEqualTo`만 사용해주셔도 괜찮을 것 같습니다.
다른 테스트들에서도 일괄적으로 수정 부탁드립니다.

na-mul replied:
사실 어떤 순서로 해주셔도 테스트 케이스는 통과하겠지만,
의미적으로 `assertThat(actual).isEqualTo(expected)`로 순서를 바꿔주셔야 할 것으로 보입니다!

na-mul commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,32 @@
+    @Test
+    void findById() {
+        Question expected = questionRepository.save(Q2);
+        Question actual = questionRepository.findById(expected.getId()).get();
+
+        assertThat(expected).isEqualTo(actual);
```
```suggestion
        Optional<Question> actual = questionRepository.findById(expected.getId());

        assertThat(actual).hasValue(expected);
```

와 같은 방식을 사용하면 Optional도 의미있게 테스트로 활용할 수 있게 됩니다. 참고 부탁드려요!

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,12 +3,34 @@
-    private ContentType contentType;
     private Long contentId;
-    private Long deletedById;
+    @Enumerated(EnumType.STRING)
+    private ContentType contentType;
```
`@Enumerated(EnumType.STRING)`도 꼼꼼히 잘 사용해주셨습니다 👍

na-mul reviewed (changes requested):
안녕하세요 대현님!
함께 JPA 미션을 진행하게 된 노동환입니다.
꼼꼼하게 진행해주신 과제 잘 확인하였습니다! 🙏👍
몇가지 코멘트를 남겨보았어요. 확인해주셔서 다시 request를 부탁드립니다!

daehyeon0 committed:
feat: DDL 삭제

daehyeon0 committed:
test(AnswerTest): 잘못된 테스트 수정

- 변수의 순서를 의미적으로 올바른 순서로 변경
- isSameAs를 이용한 테스트 삭제 (너무 엄격함)

daehyeon0 committed:
test(QuestionTest): 잘못된 테스트 수정

- 변수의 순서를 의미적으로 올바른 순서로 변경
- isSameAs를 이용한 테스트 삭제 (너무 엄격함)
- Optional 강제 unwrap하여 사용하지 않도록 변경

daehyeon0 committed:
test(UserTest): 잘못된 테스트 수정

- 변수의 순서를 의미적으로 올바른 순서로 변경
- isSameAs를 이용한 테스트 삭제 (너무 엄격함)
- Optional 강제 unwrap하여 사용하지 않도록 변경

daehyeon0 committed:
test(DeleteHistoryTest): 잘못된 테스트 수정

- 변수의 순서를 의미적으로 올바른 순서로 변경
- isSameAs를 이용한 테스트 삭제 (너무 엄격함)
- Optional 강제 unwrap하여 사용하지 않도록 변경
- 의미없는 변수 삭제

daehyeon0 committed:
refactor (Answer, Question, User): 중복되는 Entity를 BaseEntity로 추출하여 상속

daehyeon0 committed:
refactor(Answer, Question, User): 사용되지 않는 getter, setter 메서드 삭제

daehyeon0 committed:
refactor(DeleteHistoryService, QnaService): repository들을 final화

daehyeon0 committed:
refactor: DeleteHistoryService의 Transactional Propagation.REQUIRES_NEW를 REQUIRES로 변경

daehyeon0 committed:
docs: 기능 요구사항 추가

- DeleteHistoryService, QnAService 관련 요구사항 추가

daehyeon0 committed:
refactor: List<Answer>를 Answers 클래스로 변환

daehyeon0 committed:
refactor(QnaService): DeleteQuestion 리팩토링

- 일급 컬렉션: List<DeleteHistory>를 deleteHistories로 변환
- Answers.setDeleted 추가

daehyeon0 committed:
refactor: DeleteHistories 클래스 추가

- 기존 커밋에서 빼먹은것 추가

daehyeon0 commented:
단순히 엔티티 매핑만 하면 된다고 생각해버렸네요..
말씀해 주신 부분과 리팩토링 일부를 포함해 다시 요청드립니다~

질문도 하나 있는데요.
List<DeleteHistory>를 DeleteHistories로 만드는 과정에서, Question과 Answer를 컬렉션에 포함하는 메서드(add)를 오버라이드 해서 2개 만들었습니다.
하나는 Question을 받는 메서드고, 하나는 Answers를 받는 함수였는데요.
add 메서드를 앞으로도 수정하지 않으려면 Questions를 받는편이 나을거라는 생각이 드는데, 실제로 지금 의미있게 사용되는 메서드는 아니라는 부분에서 그렇게 구현을 해야하나 고민을 하다가 Question을 받는 함수로 남겨 두었습니다.
이런 경우는 보통 어떻게 하시는지 궁금해서 질문드립니다.

na-mul commented on README.md:
```diff
@@ -0,0 +1,11 @@
+- [x] DeleteHistory Entity 매핑
+- [x] Answer Entity 매핑
+- [x] DeleteHistoryService의 Transactional Propagation = REQUIRES_NEW를 제거
+- QnAService
+  - DeleteQuestion 리팩토링
+    - [x] 일급 컬렉션: List<Answer>를 Answers로 변환
+    - [x] 일급 컬렉션: List<DeleteHistory>를 deleteHistories로 변환
+    - [x] Answers.setDeleted 추가
```
1단계에서 여기까지 해주신 분은 처음인 것 같아요!
리펙토링에 정말 고생이 많으셨겠지만, 그만큼 배우는 것이 있으셨기를 바라겠습니다 👍
열정에 👏

na-mul commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,46 @@
+		this.answers = answers;
+	}
+
+	public boolean haveNotOwner(User loginUser) {
+		for (Answer answer : answers) {
+			if (!answer.isOwner(loginUser)) {
+				return true;
+			}
+		}
+		return false;
+	}
```
요 부분은 depth가 2 이상이네요! 
다음 단계에서 수정을 부탁드릴께요.

na-mul commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,13 @@
+import javax.persistence.Column;
+import javax.persistence.MappedSuperclass;
+
+@MappedSuperclass
+public class BaseEntity {
+	@Column(nullable = false)
+	protected LocalDateTime createdAt;
+	protected LocalDateTime updatedAt;
+}
```
`Auditing` 기능을 이용하여 `createdAt`과 `updatedAt`을 채워볼 수 있습니다.
https://velog.io/@conatuseus/2019-12-06-2212-%EC%9E%91%EC%84%B1%EB%90%A8-1sk3u75zo9
위 링크의 글 정도로 적용해봐도 충분할 것 같아요!

na-mul commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DeleteHistories {
+	private final List<DeleteHistory> deleteHistories = new ArrayList<>();
```
일급 컬렉션 굿! 입니다

na-mul commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,33 @@
+
+    @Test
+    void findById() {
+        Question expected = questionRepository.save(Q2);
+        Optional<Question> actual = questionRepository.findById(expected.getId());
+
+        assertThat(actual).hasValue(expected);
```
테스트도 잘 수정해주셨네요. 감사합니다!

na-mul reviewed (approved):
안녕하세요 대현님!
코멘트 드린 내용을 넘어서, 스스로 더 많은 개선점을 찾아주신 점 정말 감사드립니다.
이번 단계는 approve 하도록 하겠습니다!
다음 단계에서도 열정적인 모습 기대하겠습니다 🙇‍♂️🙏

na-mul commented:
> List를 DeleteHistories로 만드는 과정에서, Question과 Answer를 컬렉션에 포함하는 메서드(add)를 오버라이드 해서 2개 만들었습니다.
하나는 Question을 받는 메서드고, 하나는 Answers를 받는 함수였는데요.
add 메서드를 앞으로도 수정하지 않으려면 Questions를 받는편이 나을거라는 생각이 드는데, 실제로 지금 의미있게 사용되는 메서드는 아니라는 부분에서 그렇게 구현을 해야하나 고민을 하다가 Question을 받는 함수로 남겨 두었습니다.
이런 경우는 보통 어떻게 하시는지 궁금해서 질문드립니다.

음, 아마 정말 앞으로의 일을 생각한다면, `Questions`를 받아야 할 수도 있어요. 여러개의 질문을 한번에 삭제한다면 말이죠.
다만 현재 구현된 기능들, 그리고 기능의 추가 없이 리팩터링하는 정도로 본 과제가 종료될 것이 예상된다면,
`Question`만 add하도록 하여도 충분할 것입니다.

항상 앞으로 생길 모든 기능을 예상하여 어플리케이션을 개발할 필요는 없습니다.
다만 앞으로 기능 변경 사항이 들어와도 괜찮도록, 변화에 용이한 코드가 어떤 것일까 항상 고민하는 자세가 중요합니다.
지금처럼 단위테스트를 잘 작성하고, 객체지향 생활체조 원칙을 잘 지켜 나간다면 자연스럽게 그러한 코드가 만들어질 것입니다.
